compiler/src/dotty/tools/dotc/core/tasty/TreePickler.scala

```diff
@@ -332,6 +332,8 @@ class TreePickler(pickler: TastyPickler) {
 
   def pickleDef(tag: Int, mdef: MemberDef, tpt: Tree, rhs: Tree = EmptyTree, pickleParams: => Unit = ())(using Context): Unit = {
     val sym = mdef.symbol
+    if sym.isScala2MacroInScala3 then return
```

Comment:
Doesn't this break the use of Scala 3 compiled artifacts containing mixed Scala 2/3 macros from Scala 2?

If I understand this change correctly, given the following code example:

```scala
object MacroCompat {
  trait LocationMacro {
    inline implicit def location: Location = ${ locationImpl() }
    implicit def location: Location = macro MacroCompatScala2.locationImpl
  }
  def locationImpl()(using Quotes): Expr[Location] = '{ new Location(${Expr(0)}) }
}
```

this `def`:

```
implicit def location: Location = macro MacroCompatScala2.locationImpl
```

will no longer be pickled in TASTy.  But doesn't that mean that the `location` macro is now inaccessible from Scala 2 using `-Ytasty-reader`?

---

As a test, I tried a small adaptation (repo [here](https://github.com/griggt/mix-macros)) of the [mixing macros example](https://scalacenter.github.io/scala-3-migration-guide/docs/tutorials/macro-mixing.html) in the Scala 3 migration guide with this PR.  (This also required changing the TASTy minor and experimental versions to 0 to force compatibility with 2.13.6.) Compilation failed due to the missing method:

```scala
sbt:mix-macros> ++3.0.1-RC1-bin-SNAPSHOT! exampleLib/compile
  ... successful compilation ...

sbt:mix-macros> ++2.13.6 exampleTest/run
[info] Setting Scala version to 2.13.6 on 1 projects.
[info] Excluded 2 projects, run ++ 2.13.6 -v for more details.
[info] Reapplying settings...
[info] set current project to mix-macros (in build file:/src/mix-macros/)
[info] compiling 1 Scala source to /src/mix-macros/test/target/scala-2.13/classes ...
[error] /src/mix-macros/test/src/main/scala/Test.scala:5:39: Unsupported Scala 3 inline macro method location; found in object location.Macros.
[error]     println(s"We are on line ${Macros.location.line}")
[error]                                       ^
[error] one error found
```

I ran into the same issue attempting to use a version of MUnit compiled with this PR from Scala 2.13.6 (the Scala 3 published MUnit artifact is [intended to be usable from Scala 2](https://scalameta.org/munit/blog/2021/01/05/macromix.html)):

```
[error] /tmp/munit-simple/src/test/scala/Test.scala:2:23: Unsupported Scala 3 inline macro method generate; found in trait munit.internal.MacroCompat.LocationMacro.
```

```scala
// munit-simple/src/test/scala/Test.scala
class Test1 extends munit.FunSuite {
  test("Simple test") {
    assertEquals(1,2)
  }
}
```

Reply:
Thank you for looking into it @griggt . Yes, you are right, we have to pickle the Scala 2 macro definition to support the migration use case. I'll try a different fix.

Reply:
@griggt In c259f4d, we have a different fix that does not touch the pickler.